toc/rfc/rfc-draft-service-binding-rotation.md

```diff
@@ -0,0 +1,120 @@
+        "app": {
+            "data": {
+                "guid": "<myApp-guid>"
+            }
+        },
+        "service_instance": {
+            "data": {
+                "guid": "<mySerivce-guid>"
+            }
+        }
+    },
+    "type": "app"
+}'
+
+{"errors":[{"detail":"The app is already bound to the service instance","title":"CF-UnprocessableEntity","code":10008}]}
+```
+As a result, the only way to rotate service credential bindings currently is by performing blue-green deployments. However, many CF application developers are reluctant to push a new version of their application solely for the purpose of credential rotation.
+
+## Proposal
+
+The CC should allow multiple service credential bindings per service instance and application. Only the newest one should be visible in the application VCAP_SERVICES env var (or file-based service binding information) so that existing applications don’t need to change.
+
+### CF Cloud Controller
+
+#### POST /v3/service_credential_bindings
+
+Shall allow the creation of multiple service credential bindings for the same app and service instance under the following conditions:
+- service credential bindings are of type `app`
+- service instance is a managed service instances
+- service credential binding name is not changed
+
+The number of multiple service credential bindings for the same app and service instance should be limited.
+
+#### GET /v3/service_credential_bindings
+
+The API `GET /v3/service_credential_bindings` will return all service credential bindings of an application including multiple bindings to the same service instance.
+
+### Mapping of service credential bindings into application containers
+
+When creating the service credential binding information like VCAP_SERVICES or file-based one the CC must select the newest one per service instance only that is in state `succeeded`. This must be based on the `created_at` and `last_operation.state` fields from the credential binding.
+
+An application restart or restage as today will be required to activate the new binding.
+
+### CF CLI
+
+Users can create multiple service credential bindings by invoking the `bind-service` command multiple times. Example:
+```
+cf bind-service myApp myService  # initial binding
+cf bind-service myApp myService  # additional binding to the same service instance
+```
+
+`cf unbind-service myApp myService` shall delete all existing service credential bindings for `myApp` to `myService`.
+An additional parameter `cf unbind-service --guid <guid>` should support the deletion of a single service credential binding.
+
+`cf service myService` should list all bindings to apps including their `guid` and `created_at` timestamp. This information is helpful to understand which binding will be mapped into the application container.
+
+The cleanup of old service credential bindings should be supported by a new CF CLI command:
+```
+cf cleanup-outdated-service-bindings myApp [myService] [--keep-last 1]
```

Comment:
Optional CLI arguments should be flags: https://github.com/cloudfoundry/cli/wiki/CF-CLI-Style-Guide#args-vs-flags

Reply:
Changed to `cf cleanup-outdated-service-bindings myApp [--service-instance myService] [--keep-last 1]`

Very likely, the command will also get a `--wait` flag to deal with async service bindings.